### ISSUE TITLE
not-an-iterable false positive with decorator

### ISSUE THREAD

**Issue by JulienPalard** [pylint-dev/pylint#9251]
### Bug description

This reproducer uses a decorator to make a normal function an infinite iterable, sadly pylint don't get it:

```python
"""repro.py file"""

from itertools import count


class IteratorFromFunction: # pylint: disable=too-few-public-methods
    """Builds an iterator based on a function."""

    def __init__(self, function):
        """Build a new sequence."""
        self.function = function

    def __iter__(self):
        """Iterate over the integer sequence."""
        for i in count():
            yield self.function(i)


@IteratorFromFunction
def i_am_iterable(i):
    """Sequence of square numbers."""
    return i ** 2


def main():
    """just to check it works"""
    for i in i_am_iterable:
        print(i)
        if i > 100:
            break

if __name__ == "__main__":
    main()
```

### Command used

```shell
pylint repro.py
```

### Pylint output

```python
$ pylint repro.py
************* Module repro
repro.py:27:13: E1133: Non-iterable value i_am_iterable is used in an iterating context (not-an-iterable)

------------------------------------------------------------------
Your code has been rated at 6.88/10 (previous run: 6.88/10, +0.00)
```

### Expected behavior

The decorator actually replace the function with an iterable, so pylint should not tell it's not.

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.6 (main, Oct  8 2023, 05:06:43) [GCC 13.2.0]
```

### OS / Environment

Debian Trixie
